### DISCUSSION TITLE
Ideas for tests!

### DISCUSSION THREAD

**Discussion by m-col** [qtile/qtile#2134] [Ideas]
I was going to add an issue for this but discussions looks like it is aimed at this kind of thing, and it looks quite nice!

Every now and then I think of something we should test (often due to a newly introduced bug from the previous commit that wasn't caught by the CI). I usually don't get round to writing the test immediately, so I figured it might be helpful to have a place we can keep a list of tests we could/should write to serve as a bit of a to do list. It can also serve as a list of small, easy tasks that people who want to contribute but are usure where to start can work on.

Feel free to comment in disagreement at any ideas, edit this comment to add your own, or tick them off if you're going to work on it.

Ideas:

- [x] Reorganise files and folders to more closely mirror libqtile (IMO this would be more tidy and organised but is less important)
- [ ] Test widget mouse callbacks by emulating clicks
- [ ] Test window move/resize with emulated drags
- [x] Add coveralls to CI #2054 
- [ ] Test `_configure` at the various class levels, particularly when it is called more than once
- [ ] use `shutils.where` instead of the `whereis` function in `conftest.py`
- [ ] test transparency
- [ ] test wayland core, mirror x11 tests
- [x] vertical bars and widgets that support them #3428
- [ ] hooks could do with a look to over ensure they are all tested
- [ ] test fullscreening behaviour

**Comment by elParaguayo** [comment]
Having been responsible for a bug that's now in 0.17.0 (see #2217) I think some basic tests for widgets would be an excellent idea...

**Reply by m-col** [reply]
Feel free to edit any specific to-dos into the bullet list at the top if that is helpful at all :)

**Comment by elParaguayo** [comment]
How basic can tests be? 

For example, the backlight widget relies on reading a file for all of its functionality so we could probably only test if it initialises ok. That's probably worth having as it would catch changes that completely break the widget.

**Reply by tych0** [reply]
I think anything that increases coverage would help. I definitely don't think we need to get to 100% test coverage, but something that does some kind of basic smoke test of the widgets with their default configuration would be appreciated.

**Comment by elParaguayo** [comment]
Agreed. Am drafting something that initialises the widget and overrides a default value to check they load ok and set the values.

**Reply by tych0** [reply]
I'd also _configure() them too, just so that we can get the timers running too.

**Reply by elParaguayo** [reply]
OK - will take a look at that too. Quite a few of them do things in `_configure` that's a bit harder to test.

**Comment by elParaguayo** [comment]
I'm working my way through writing tests for the widgets. One thing that's come up a few times is that I can't test logging output. The reason I'd like to do this is that, for some widgets, a log entry is the only evidence that some part of the code has run.

From this section of `test\conftest.py`, it looks like the `log_path` is set to `None` which I'm assuming means there's no log.
https://github.com/qtile/qtile/blob/569c4d9aaad1dbd912435648f5f814e084de8365/test/conftest.py#L280-L297

Is there a reason for this? Performance? Not a valuable test? (I note that `test_utils` doesn't test the logger).

It's not a big deal but was just curious about the rationale.

EDIT: I suspect I'm completely wrong about the above. If I call `pytest` directly on my machine then `caplog` works so I suspect there's some other configuration point that I'm missing.

**Reply by m-col** [reply]
I think the `logging` module can let you specify an arbitrary function to use as a callback, which might be useful for capturing log output during tests

**Comment by m-col** [comment]
I'm planning to do some refactoring of the test suite with the goal of testing out the wayland backend in the same way as the x11 backend and keeping things nice and tidy while doing so. Before I start though it might be best to spitball on what the final form might look like.

One idea would be to run (almost) all tests using both backends. Many tests use xcffib or x11 `Window` specific code to test functionality and so many tests are inherently x11-specific in their current form - writing a wayland equivalent would likely be the way to go, rather than use the same test for two backends. For this I'm picturing some default fixture parameterising that would use one backend and then the other, but I expect implementing this would be tricky.

Another idea would be to split things that are being tested in such a way that all tests that directly involve backend logic are contained within the respective backend's test/backend/folder. For other tests that test other parts of Qtile that do not directly involve a backend could instead use a test `Core` that would either no-op or assert things to make sure state makes sense. This approach would likely be more performant (especially vs the above idea) but then tests for non-backend-specific things such as the scratchpads would not be tested across different cores. I'm not sure how important this would be - the assumption would be that if the scratchpad (in this example) works fine with a test backend, and the x11 and wayland backends behave as expected, then they would work as expected together.

Anyone have any thoughts as to which they think would be better? I myself am leaning towards the latter. Or does anyone have any alternative suggestions for how things might look?

**Reply by elParaguayo** [reply]
How important is performance? I would think we should pick a more comprehensive test suite over a more performant one. The bit I can't quite gauge is how many of the tests would result in needless duplication.

Still, if there's any easy way to parametrise test to use both backends, that's probably the way I'd go if I were doing it.

**Reply by m-col** [reply]
Performance is not super important, just a convenience for development really.

Yeah I guess you're right - it might be a bit too optimistic to think that we wouldn never come across a random bug in some apparently backend-unrelated code that only happens when using a specific backend :laughing:
